drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,344 @@
+#endif
+	};
+
+	memcpy(buf, &hdr, sizeof(hdr));
+	read_buf = buf + sizeof(hdr);
+	buf_avail = buf_len - sizeof(hdr);
+
+	/*
+	 * Prepare rtio enabled bus to read all fifo_count entries from
+	 * LSM6DSV16X_FIFO_DATA_OUT_TAG.  Then lsm6dsv16x_complete_op_cb
+	 * callback will be invoked.
+	 *
+	 * STMEMSC API equivalent code:
+	 *
+	 *   num = fifo_status.fifo_level;
+	 *
+	 *   while (num--) {
+	 *     lsm6dsv16x_fifo_out_raw_t f_data;
+	 *
+	 *     lsm6dsv16x_fifo_out_raw_get(&dev_ctx, &f_data);
+	 *   }
+	 */
+	struct rtio_sqe *write_fifo_dout_addr = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	struct rtio_sqe *read_fifo_dout_reg = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	struct rtio_sqe *complete_op = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	uint8_t reg = lsm6dsv16x_bus_reg(lsm6dsv16x, LSM6DSV16X_FIFO_DATA_OUT_TAG);
+
+	rtio_sqe_prep_tiny_write(write_fifo_dout_addr, iodev, RTIO_PRIO_NORM, &reg, 1, NULL);
+	write_fifo_dout_addr->flags = RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_read(read_fifo_dout_reg, iodev, RTIO_PRIO_NORM,
+			   read_buf, buf_avail, lsm6dsv16x->streaming_sqe);
+	read_fifo_dout_reg->flags = RTIO_SQE_CHAINED;
+	if (lsm6dsv16x->bus_type == 0) {
+		read_fifo_dout_reg->iodev_flags |= RTIO_IODEV_I2C_STOP | RTIO_IODEV_I2C_RESTART;
+	}
+	rtio_sqe_prep_callback(complete_op,
+			       lsm6dsv16x_complete_op_cb, (void *)dev, lsm6dsv16x->streaming_sqe);
+
+	rtio_submit(lsm6dsv16x->rtio_ctx, 0);
+}
+
+void lsm6dsv16x_stream_irq_handler(const struct device *dev)
+{
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+	struct rtio_iodev *iodev = lsm6dsv16x->iodev;
+
+	/* proceed only if streaming */
+	if (lsm6dsv16x->streaming_sqe == NULL) {
+		return;
+	}
+
+	lsm6dsv16x->fifo_status[0] = lsm6dsv16x->fifo_status[1] = 0;
+
+	/*
+	 * Prepare rtio enabled bus to read LSM6DSV16X_FIFO_STATUS1 and LSM6DSV16X_FIFO_STATUS2
+	 * registers where FIFO threshold condition and count are reported.
+	 * Then lsm6dsv16x_read_fifo_cb callback will be invoked.
+	 *
+	 * STMEMSC API equivalent code:
+	 *
+	 *   lsm6dsv16x_fifo_status_t fifo_status;
+	 *
+	 *   lsm6dsv16x_fifo_status_get(&dev_ctx, &fifo_status);
+	 */
+	struct rtio_sqe *write_fifo_status_addr = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	struct rtio_sqe *read_fifo_status_reg = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	struct rtio_sqe *check_fifo_status_reg = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+	uint8_t reg = lsm6dsv16x_bus_reg(lsm6dsv16x, LSM6DSV16X_FIFO_STATUS1);
+
+	rtio_sqe_prep_tiny_write(write_fifo_status_addr, iodev, RTIO_PRIO_NORM, &reg, 1, NULL);
+	write_fifo_status_addr->flags = RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_read(read_fifo_status_reg, iodev, RTIO_PRIO_NORM,
+			   lsm6dsv16x->fifo_status, 2, NULL);
+	read_fifo_status_reg->flags = RTIO_SQE_CHAINED;
+	if (lsm6dsv16x->bus_type == 0) {
+		read_fifo_status_reg->iodev_flags |= RTIO_IODEV_I2C_STOP | RTIO_IODEV_I2C_RESTART;
+	}
+	rtio_sqe_prep_callback(check_fifo_status_reg,
+			       lsm6dsv16x_read_fifo_cb, (void *)dev, NULL);
```

Comment:
Not required, but one possibility here is to look at the current data request option and change the callback here, avoiding some branching potentially in the fifo cb.

Potentially you could directly read the FIFO perhaps because the watermark is set, you can see if the data is requested with it.

Other possibilities to reduce branching are likely here but I get the current example drivers don't do this.

Reply:
probably worth doing...

```c
check_fifo_status_reg->flags |= RTIO_SQE_NO_RESPONSE;
```

Reply:
I had forgotten @JordanYates added a helper for this,

rtio_sqe_prep_callback_no_cqe()

 https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/rtio/rtio.h#L618